pkg/s3select/sql/record.go

```diff
@@ -0,0 +1,198 @@
+	alias        string
+	indexNameMap map[int64]string
+	nameValueMap map[string]*Value
+	index        int64
+}
+
+// String - returns string representation of this record.
+func (r *Record) String() string {
+	var items []string
+	r.Range(func(name string, value *Value) bool {
+		items = append(items, name+":"+fmt.Sprintf("%v", value))
+		return true
+	})
+	return fmt.Sprintf("{%v}", strings.Join(items, ", "))
+}
+
+// MarshalCSV - encodes to CSV data.
+func (r *Record) MarshalCSV(fieldDelimiter rune) ([]byte, error) {
+	var csvRecord []string
+	r.Range(func(name string, value *Value) bool {
+		csvRecord = append(csvRecord, value.CSVString())
+		return true
+	})
+
+	buf := new(bytes.Buffer)
+	w := csv.NewWriter(buf)
+	w.Comma = fieldDelimiter
+	if err := w.Write(csvRecord); err != nil {
+		return nil, err
+	}
+	w.Flush()
+	if err := w.Error(); err != nil {
+		return nil, err
+	}
+
+	data := buf.Bytes()
+	return data[:len(data)-1], nil
+}
+
+// MarshalJSON - encodes to JSON data.
+func (r *Record) MarshalJSON() ([]byte, error) {
+	rv := make(map[string]interface{})
+	for name, value := range r.nameValueMap {
+		if r.alias != "" && strings.HasPrefix(name, r.alias+".") {
+			name = strings.TrimPrefix(name, r.alias+".")
+		}
+
+		tokens := strings.Split(name, ".")
+
+		if len(tokens) == 1 {
+			rv[tokens[0]] = value.Value()
+			continue
+		}
+
+		var leaf map[string]interface{}
+		for i := range tokens[:len(tokens)-1] {
+			if leafValue, ok := rv[tokens[i]]; ok {
+				leaf = leafValue.(map[string]interface{})
+			} else {
+				leaf = make(map[string]interface{})
+				rv[tokens[i]] = leaf
+			}
+		}
+
+		leaf[tokens[len(tokens)-1]] = value.Value()
+	}
+
+	return json.Marshal(rv)
+}
+
+// UnmarshalCSV - decodes CSV data to record.
+func (r *Record) UnmarshalCSV(columnNames, csvRecord []string) error {
+	if len(columnNames) != len(csvRecord) {
+		return errMissingHeaders(fmt.Errorf("column names and csv record do not match"))
+	}
+
+	for i, name := range columnNames {
+		r.set(name, NewString(csvRecord[i]))
+	}
+
+	return nil
+}
+
+func (r *Record) populate(nameValueMap map[string]interface{}, prefix string) error {
+	for name, value := range nameValueMap {
+		if prefix != "" {
+			name = prefix + "." + name
+		}
+
+		switch value.(type) {
+		case bool:
+			r.set(name, NewBool(value.(bool)))
+		case float64:
+			r.set(name, NewFloat(value.(float64)))
+		case string:
+			r.set(name, NewString(value.(string)))
+		case map[string]interface{}:
+			if err := r.populate(value.(map[string]interface{}), name); err != nil {
+				return err
+			}
+		default:
+			return errUnrecognizedFormatException(fmt.Errorf("unsupported value type %T; %v", value, value))
+		}
+	}
+
+	return nil
+}
+
+// UnmarshalJSON - decodes JSON data to record.
+func (r *Record) UnmarshalJSON(data []byte) error {
+	nameValueMap := make(map[string]interface{})
+	if err := json.Unmarshal(data, &nameValueMap); err != nil {
+		return err
+	}
+
+	return r.populate(nameValueMap, r.alias)
```

Comment:
This is losing the ordering of the JSON keys from the input data - it will lead to some inconsistent results compared to S3: for `*` expressions, S3 maintains the input ordering of the columns/fields. Maybe we can add some special handling for `*` expressions to return the data from source without re-ordering later?

Reply:
`json.Unmarshal([]byte, &map)` does not work with ordering, hence its not possible to save order.

However order is not important for JSON objects.  When `*` is used for object having JSON records, CSV output serialization is not supported. It throws an error now.

To support this case, its required to write our own JSON parser or a parser preserve ordering.

Reply:
Ordering can be maintained using sjson and gjson which is what the previous code actually did @balamurugana where you don't really have to use json.Unmarshal because json umarshalling is simply a very expensive operation even with a `map[string]interface{}` - this, in turn, requires reflection which is not needed. Since sjson allows for understanding the type properly without the use of reflection.

Reply:
Cool. I will try `sjson/gjson`.

Reply:
Even JSON input to JSON output will end up getting reordered and not looking correct if compared with S3. The only way different S3Select implementations (e.g. Minio vs S3) can be consistent is if they keep the input order of columns in the data. 

Even for Parquet input the columns have the same order in each chunk of rows, and I think S3 would use this order of the columns in the the `*` output.

IMO we can fix it in subsequent work, noting that we are introducing a regression here.

Reply:
Remove this constraint.